### DISCUSSION TITLE
静的API <CRE_MPF> が unknown static API エラーで使用できない

### DISCUSSION THREAD

**Discussion by sengokuuu** [toppers/users-forum#90] [ET-Robocon]
メモリプールを用いたプログラムを書いていましたが、ビルドするタイミングで添付画像の通りエラーが出ています。
文法ミスも疑いましたが、自分では見つけることができませんでした。
コンフィグファイルのコードを下記に記述します。ご指摘あればよろしくお願いします。

```
INCLUDE("app_common.cfg");

#include "app.h"

DOMAIN(TDOM_APP) {
CRE_TSK(MAIN_TASK, { TA_ACT, 0, main_task, TMIN_APP_TPRI + 1, STACK_SIZE, NULL });
CRE_TSK(RUN_TASK, {TA_NULL, 0, run_task, TMIN_APP_TPRI + 1, STACK_SIZE, NULL});
CRE_TSK(JUDGE_TASK, {TA_NULL, 0, judge_task, TMIN_APP_TPRI + 1, STACK_SIZE, NULL});
CRE_TSK(MANAGER_TASK, {TA_NULL, 0, manager_task, TMIN_APP_TPRI + 1, STACK_SIZE, NULL});

CRE_CYC(RUN_CYCLIC_HANDLER, {TA_NULL, 0, runCyclicHandler, 10 * 1000U, 0U});
CRE_CYC(JUDGE_CYCLIC_HANDLER, {TA_NULL, 0, judgeCyclicHandler, 10 * 1000U, 0U});
CRE_CYC(MANAGER_CYCLIC_HANDLER, {TA_NULL, 0, managerCyclicHandler, 10 * 1000U, 0U});

CRE_DTQ(SCENE_DTQ, {TA_NULL, 2, TA_NULL});
CRE_DTQ(JUDGE_DTQ, {TA_NULL, 2, TA_NULL});
CRE_DTQ(JUDGE_SENSE_DTQ, {TA_NULL, 2, TA_NULL});
CRE_DTQ(RUN_DTQ, {TA_NULL, 2, TA_NULL});
CRE_DTQ(RUN_SENSE_DTQ, {TA_NULL, 2, TA_NULL});
CRE_DTQ(MANAGER_DTQ, {TA_NULL, 2, TA_NULL});

CRE_MPF(MSG_MPF, {TA_NULL, 12, 64, NULL, NULL});

}
```

![132900](https://user-images.githubusercontent.com/109186683/189574245-9c8b89ff-7a97-4e06-b0ed-2b2a91fbd3a9.png)

**Comment by hiroakitakada** [comment]
ダウンロードするモジュールをビルドする際の話しだと思いますが，ダウンロードモジュール側では，CRE_MPFは使用できないものと思います。tlsfライブラリが使えると思うので，そちらを使って頂くのが良いと思います。

**Reply by sengokuuu** [reply]
お忙しいところありがとうございます。ダウンロードモジュール側から利用できないとのことで理解しました。TLSFライブラリというヒントも感謝いたします。

後学のために教えていただきたいのですが、CRE_TSK や CRE_DTQ といった静的APIが利用できるのに対して CRE_MPF が使えないというのはどういった理由なのでしょうか。

よろしくお願いします。

**Reply by YixiaoLi** [reply]
> CRE_TSK や CRE_DTQ といった静的APIが利用できるのに対して CRE_MPF が使えないというのはどういった理由なのでしょうか。

特に深い理由はありませんが、EV3RTの設計方針として、
① make imgでアプリをuImageとしてビルドする場合、全ての静的APIを利用できる。
② make appでアプリを動的ダウンロードできるモジュールとしてビルドする場合、ユーザアプリの開発に必要な静的APIだけサポートする。

EV3RTはTLSFライブラリを用いてC/C++標準の動的メモリ確保関数（malloc/free, new/delete）をサポートしています。そのため、ユーザアプリはCRE_MPFを使わなくても良いと判断し、動的ダウンロード機能でこのAPIのサポートを実装していません。

**Reply by sengokuuu** [reply]
理解しました。お忙しいところありがとうございました。
